src/_pytest/monkeypatch.py

```diff
@@ -216,7 +216,7 @@ def setenv(self, name, value, prepend=None):
         value = str(value)
         if prepend and name in os.environ:
             value = value + prepend + os.environ[name]
-        self.setitem(os.environ, name, value)
+        self.setitem(os.environ, str(name), value)
```

Comment:
I'm hesitant because this makes it less than a simple wrapper around `os.environ.__setitem__` -- blind `str` is also often a bug.

For instance this now turns `.setenv(b'foo', b'bar')` in python3 from a TypeError (as it should be!) to settings `os.environ["b'foo'"] = "b'bar'"`

This was already (imo) a problem for `.setenv('foo', b'bar')` so it's not exactly a regression in that manner

Reply:
You got a point, probably makes sense to only call `str()` on Python 2 then.

Reply:
imo pytest shouldn't be touching these parameters at all and just passing through to `os.environ`, any behaviour beyond that seems against "principle of least surprise".  it is unfortunate that modifications here can lead to latent errors in subprocess but paving those over in pytest seems more likely to cause bugs than fix them 🤷‍♂️

Reply:
line 216 covers the value

Reply:
Not sure, I think we can "do the right thing" here if we want to help users. Letting users set `unicode` objects as keys in `os.environ` is an error, which unfortunately Python does not catch early but will cause problems later on.

Reply:
yeah it is tough :/ I guess I'd be more _okay_ if we gate this with something like:

```python
if six.PY2 and windows:  # maybe even just `if PY2:`, your call!
    name, val = _to_bytes(name), _to_bytes(val)
...
```

where `_to_bytes` is something like

```python
def _to_bytes(s):
    return s if isinstance(s, bytes) else s.encode('UTF-8')
```

Reply:
@nicoddemus im wondering, what if we just triggered a warning on incompatible string types but still let users do whatever they want

Reply:
Well that's the safest approach, we will be sure to not break anything then. We should check `if not isinstance(key, str):` and it would be a valid warning for Python 2 and 3. What do you think @asottile?

Reply:
that actually sounds great 👍 

in python3 they'll get a warning and then an exception but that's probably fine!